File: LeetCode_2023/leetcode_problems/34. 在排序数组中查找元素的第一个和最后一个位置.py
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:

        res = [-1, -1]

        if not nums:
            return res
        # nums 是一个非递减数组
        l, r = 0, len(nums) - 1

        while l < r:
            mid = l + (r - l) // 2
            if target > nums[mid]:
                l = mid + 1
            else:
                r = mid

        if nums[l] == target:
            res[0] = l
        else:
            return res

        l, r = 0, len(nums) - 1

        while l < r:
            mid = l + (r - l + 1) // 2
            if target < nums[mid]:
                r = mid - 1
            else:
                l = mid
        res[1] = r

        return res


solution = Solution()
nums = [5, 7, 7, 8, 8, 10]
target = 8
print(solution.searchRange(nums, target))

File: example_c++/34.在排序数组中查找元素的第一个和最后一个位置.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        vector<int> result = {-1, -1};

        if (nums.size() == 0) {
            return result;
        }

        int left = 0;
        int right = nums.size() - 1;
        int mid;


        while (left < right) {
            mid = left + (right - left) / 2;
            if (target > nums[mid]) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        if (nums[left] == target) {
            result[0] = left;
        } else {
            return result;
        }

        left = 0;
        right = nums.size() - 1;

        while (left < right) {
            mid = left + (right - left + 1) / 2;
            if (target < nums[mid]) {
                right = mid - 1;
            } else {
                left = mid;
            }
        }

        if (nums[right] == target) {
            result[1] = right;
        }

        return result;
    }
};

int main() {
    int target = 8;
    vector<int> nums = {5, 7, 7, 8, 8, 10};
    vector<int> result = Solution().searchRange(nums, target);
    for (int res: result) {
        cout << "result: " << res << endl;
    }

    return 0;
}